libr/debug/dsession.c

```diff
@@ -231,3 +235,375 @@ R_API bool r_debug_session_add_mem_change(RDebugSession *session, ut64 addr, ut8
+	if (!sdb_sync (db)) {
+		eprintf ("Failed to sync session to %s\n", file);
+		goto end;
+	}
+	ret = true;
+end:
+	sdb_close (db);
+	sdb_free (db);
+	return ret;
+}
+
+#define CNUM(x) sdb_fmt ("%s.a[%u].cnum", addr, x)
+#define DATA(x) sdb_fmt ("%s.a[%u].data", addr, x)
+
+// Extract ut64 value from sdb_json str
+static ut64 _json_num_get(Sdb *db, const char *k, const char *p, ut32 *cas) {
+	char *vstr = sdb_json_get (db, k, p, cas);
+	if (vstr) {
+		ut64 v = sdb_atoi (vstr);
+		free (vstr);
+		return v;
+	}
+	return 0;
+}
+
+static bool load_memory(Sdb *db, HtUP *memory) {
+	size_t i, size;
+	char *addr;
+	// Iterate over "memory" items
+	char *a = sdb_get (db, "memorylist", 0);
+	sdb_aforeach (addr, a) {
+		// Insert a new vector into `memory` HtUP at `addr`
+		RVector *vmem = r_vector_new (sizeof (RDebugChangeMem), NULL, NULL);
+		if (!vmem) {
+			eprintf ("Error: failed to allocate RVector vmem.\n");
+			return false;
+		}
+		ht_up_insert (memory, sdb_atoi (addr), vmem);
+
+		// Extract <RDebugChangeMem>'s into the new vector
+		size = _json_num_get (db, "memory", sdb_fmt ("%s.size", addr, i), 0);
+		for (i = 0; i < size; i++) {
+			int cnum = _json_num_get (db, "memory", CNUM (i), 0);
+			ut8 data = _json_num_get (db, "memory", DATA (i), 0);
+			RDebugChangeMem mem = { cnum, data };
+			r_vector_push (vmem, &mem);
+		}
+		sdb_aforeach_next (addr);
+	}
+	return true;
+}
+
+static bool load_registers(Sdb *db, HtUP *registers) {
+	size_t i, size;
+	char *addr;
+	// Iterate over "registers" items
+	char *a = sdb_get (db, "reglist", 0);
+	sdb_aforeach (addr, a) {
+		// Insert a new vector into `registers` HtUP at `addr`
+		RVector *vreg = r_vector_new (sizeof (RDebugChangeReg), NULL, NULL);
+		if (!vreg) {
+			eprintf ("Error: failed to allocate RVector vreg.\n");
+			return false;
+		}
+		ht_up_insert (registers, sdb_atoi (addr), vreg);
+
+		// Extract <RDebugChangeReg>'s into the new vector
+		size = _json_num_get (db, "registers", sdb_fmt ("%s.size", addr, i), 0);
+		for (i = 0; i < size; i++) {
+			int cnum = _json_num_get (db, "registers", CNUM (i), 0);
+			ut64 data = _json_num_get (db, "registers", DATA (i), 0);
+			RDebugChangeReg reg = { cnum, data };
+			r_vector_push (vreg, &reg);
+		}
+		sdb_aforeach_next (addr);
+	}
+	return true;
+}
+
+#define SNAPATTR(ATTR) sdb_fmt ("%s.snaps.a[%u]." #ATTR, cnum, i)
+#define REGATTR(ATTR) sdb_fmt ("%s.registers.%d." #ATTR, cnum, i)
+
+static bool load_checkpoints(Sdb *db, RDebug *dbg, RVector *checkpoints) {
+	size_t i, size;
+	char *cnum;
+	// Iterate over "checkpoints" items
+	char *a = sdb_get (db, "chkptlist", 0);
+	sdb_aforeach (cnum, a) {
+		RDebugCheckpoint checkpoint = { 0 };
+		checkpoint.cnum = (int)sdb_atoi (cnum);
+
+		// Extract RRegArena's from "registers"
+		for (i = 0; i < R_REG_TYPE_LAST; i++) {
+			int size = _json_num_get (db, "checkpoints", REGATTR (size), 0);
+			if (size == 0) {
+				continue;
+			}
+			char *edata = sdb_json_get (db, "checkpoints", REGATTR (bytes), 0);
```

Comment:
`sdb_json_get` is extremely inefficient, it re-parses the entire json on every call so here you would have O(n^2) parsing. After the format change mentioned above, it will probably be O(n) so as a temporary solution `sdb_json` might be ok until nxjson is available in r2.